Temperament lookup table

riyalohia opened the pull request:
As per our discussion, I have created the lookup table with ratios but I've not written the code that send the notes to the synth for temperaments other than equal because we need to send the frequency to the synth. For frequency, we need both ratios and starting pitch. We have the ratio but starting pitch will be defined by the user. So unless we get the starting pitch we cannot send other temperaments notes to synth.
Please let me know if I am missing anything.

walterbender commented:
I will take a look. Regarding the starting pitch, it would be determined by the mode and key. By default, we are running C Major. But I suppose that every time the user changes key, we'd want to regenerate the table. @pikurasa  does that make sense?

walterbender commented:
Not sure why you made the changes to logo.js like you did. Can you explain why you test for temperament === 'equal' throughout?

riyalohia commented:
@walterbender, I agree to the starting pitch point.
Regarding changes in the logo.js,  we have to pass default note to the synth for equal temperament but at some point when we are in a different tuning system, then we need to pass frequencies to the synth. That's why I've tested for equal temperament throughout. When we are in  tuning systems, I'll include the code that will test for the systems other than equal.

riyalohia commented:
I have added 'set temperament' block along with a selector block. The pitches/notes blocks after this flow block will be tuned according to the selected temperament. I have added getFrequency function in musicutils.js which will calculate the frequency that we need to send to the synth. Also, I have added comments in getFrequency function.

I am stuck at some points:
1. Selecting custom from the selector block will give an error because we dont have custom in our lookup table.
2. I think Diminshed 2 is for the C to Dbb but I couldn't find the name for interval between C to C##. Thats why for double sharps and double flats, I have written the code to calculate frequency assuming that C## = D, Dbb = C and so on.

walterbender commented on js/utils/musicutils.js:
```diff
@@ -1233,6 +1241,83 @@ function getInterval (interval, keySignature, pitch) {
+    var t = TEMPERAMENT[temperament].ratios;
+
+    //For calculating frequency of perfect 8.
+    for (var ratio in t ) {
```
Is this loop needed? Why not simply: perfect8 = t['pefect8'];

walterbender commented on js/utils/musicutils.js:
```diff
@@ -373,6 +373,79 @@ var OSCTYPES = [
 
+const TEMPERAMENT = {
+    'equal': {
+        'ratios': {
```
Do we need to define the intervals inside of a ratios object? Why the extra layer?
e.g., TEMPERAMENT['equal']['ratios']['unison'] could just be TEMPERAMENT['equal']['unison'].

walterbender commented on js/logo.js:
```diff
@@ -257,6 +258,7 @@ function Logo () {
     this.neighborArgNote2 = {};
     this.neighborArgBeat = {};
     this.neighborArgCurrentBeat = {};
+    this.frequency;
```
does this need to be an array so we can have chords? And does it need to be one per turtle for multiple voices?

riyalohia commented:
@walterbender , As per your comment, I have created a lookup table in synth utils to map all the note names into frequency each time temperament is changed.

walterbender commented:
It basically seems to work, although I am seeing no difference between meantone and just intonation when I print out frequencies. Not sure why yet,

We can worry about cleaning up the code once we confirm everything is working as expected (right now, the calculation happens each time a note is played, which is not necessary.)

@pikurasa  can you test as well?

walterbender commented:
<pre>
equal
C4
261.6255653005986

equal
C4
261.6255653005986

equal
D4
293.6647679174076

equal
C4
261.6255653005986

equal
D4
293.6647679174076

equal
E4
329.6275569128699

equal
F4
349.2282314330039

equal
C4
261.6255653005986

equal
D4
293.6647679174076

equal
E4
329.6275569128699

equal
F4
349.2282314330039

equal
G4
391.99543598174927

equal
A4
439.99999999999994

equal
B4
493.88330125612407

equal
C5
523.2511306011972

equal
C4
261.6255653005986

equal
D4
293.6647679174076

equal
E4
329.6275569128699

equal
F4
349.2282314330039

equal
G4
391.99543598174927

equal
A4
439.99999999999994

equal
B4
493.88330125612407

equal
C5
523.2511306011972

equal
D5
587.3295358348151

equal
E5
659.2551138257398

equal
F5
698.4564628660078

equal
G5
783.9908719634985

equal
A5
879.9999999999999

equal
B5
987.7666025122481

equal
C6
1046.5022612023945

equal
D6
1174.6590716696303

equal
C4
261.6255653005986

equal
D4
293.6647679174076

equal
E4
329.6275569128699

equal
F4
349.2282314330039

equal
G4
391.99543598174927

equal
A4
439.99999999999994

equal
B4
493.88330125612407

equal
C5
523.2511306011972

equal
D5
587.3295358348151

equal
E5
659.2551138257398

equal
F5
698.4564628660078

equal
G5
783.9908719634985

equal
A5
879.9999999999999

equal
B5
987.7666025122481

equal
C6
1046.5022612023945

equal
D6
1174.6590716696303

equal
E6
1318.5102276514797

equal
F6
1396.9129257320155

equal
G6
1567.981743926997

equal
A6
1759.9999999999998

equal
B6
1975.5332050244963

equal
C7
2093.004522404789

equal
D7
2349.3181433392606

equal
E7
2637.0204553029594

equal
F7
2793.825851464031

equal
G7
3135.963487853994

equal
A7
3519.9999999999995

equal
B7
3951.0664100489926

equal
C8
4186.009044809578

equal
D8
4698.636286678521

equal
E8
5274.040910605919

equal
F8
5587.651702928062

meantone
C4
261.6255653005986

meantone
C4
261.6255653005986

meantone
D4
294.32876096317347

meantone
C4
261.6255653005986

meantone
D4
294.32876096317347

meantone
E4
327.0319566257483

meantone
F4
348.8340870674648

meantone
C4
261.6255653005986

meantone
D4
294.32876096317347

meantone
E4
327.0319566257483

meantone
F4
348.8340870674648

meantone
G4
392.43834795089793

meantone
A4
436.0426088343311

meantone
B4
490.5479349386224

meantone
C5
523.2511306011972

meantone
C4
261.6255653005986

meantone
D4
294.32876096317347

meantone
E4
327.0319566257483

meantone
F4
348.8340870674648

meantone
G4
392.43834795089793

meantone
A4
436.0426088343311

meantone
B4
490.5479349386224

meantone
C5
523.2511306011972

meantone
D5
588.6575219263469

meantone
E5
654.0639132514966

meantone
F5
697.6681741349296

meantone
G5
784.8766959017959

meantone
A5
872.0852176686622

meantone
B5
981.0958698772448

meantone
C6
1046.5022612023945

meantone
D6
1177.3150438526939

meantone
C4
261.6255653005986

meantone
D4
294.32876096317347

meantone
E4
327.0319566257483

meantone
F4
348.8340870674648

meantone
G4
392.43834795089793

meantone
A4
436.0426088343311

meantone
B4
490.5479349386224

meantone
C5
523.2511306011972

meantone
D5
588.6575219263469

meantone
E5
654.0639132514966

meantone
F5
697.6681741349296

meantone
G5
784.8766959017959

meantone
A5
872.0852176686622

meantone
B5
981.0958698772448

meantone
C6
1046.5022612023945

meantone
D6
1177.3150438526939

meantone
E6
1308.1278265029932

meantone
F6
1395.3363482698592

meantone
G6
1569.7533918035917

meantone
A6
1744.1704353373243

meantone
B6
1962.1917397544896

meantone
C7
2093.004522404789

meantone
D7
2354.6300877053877

meantone
E7
2616.2556530059865

meantone
F7
2790.6726965397183

meantone
G7
3139.5067836071835

meantone
A7
3488.3408706746486

meantone
B7
3924.3834795089792

meantone
C8
4186.009044809578

meantone
D8
4709.260175410775

meantone
E8
5232.511306011973

meantone
F8
5581.345393079437
</pre>

pikurasa commented:
> @pikurasa can you test as well?

Correct me if I am wrong, but I do not see anything in the UI that I can test yet. Riya, when you have a chance, please put something in the UI for me to test, even if it is very simple? Maybe just the block with the dropdown of choices for temperament (the one that triggers a change in the tuning of notes for subsequent pitches)

walterbender commented:
@pikurasa  There is a set temperament block that lets you select between equal, meantone, and just intonation

pikurasa commented:
I found it!

I was on the wrong branch... sometimes git eludes me...

pikurasa commented:
It seems to work.

I need to test again when I can turn up the volume and hopefully on nicer speakers as well.

This is a very, very simple test file.
[2018-05-18-Temperament-Test.html.zip](https://github.com/sugarlabs/musicblocks/files/2019045/2018-05-18-Temperament-Test.html.zip)

More sophisticated, robust test. 
[2018-05-18-Temperament-Test-12-keys.html.zip](https://github.com/sugarlabs/musicblocks/files/2019052/2018-05-18-Temperament-Test-12-keys.html.zip)


...(my listening circumstances are bad, but) I think I hear more interference for Just Intonation than I would expect. ...Yes, in particularly, for the `I` chord in 2nd example. Is the math correct for this? I will investigate which particular intervals sound bad later--my hunch is C <--> B (a seventh away)

Also, the attack is not smooth (unsure if this is related to recent changes--perhaps performance-related). The attack is always going "ba-dum, ba-dum" instead of starting immediately.

riyalohia commented:
@walterbender @pikurasa , Do I need to make changes in calculations?

walterbender commented:
I am still not seeing (or hearing) any difference between meantone and just intonation. I'll dig a little deeper.

pikurasa commented:
@riyalohia Please investigate to make sure that your new calculations are being run. I am listening to it again and I do not notice much of a difference in the sound. Just Intonation, for example, will have a very "pure" sound for the chosen key as it mathematically avoids interference in the resultant sound waves.

pikurasa commented:
@riyalohia I suggest this, which we will want anyway, but will also help with debugging. Please add a block that shows the "pitch in hertz"

This is a quick mockup for the general idea:

![screenshot at 2018-05-21 15 45 21](https://user-images.githubusercontent.com/13454579/40326660-1725a604-5d0e-11e8-8ec5-949a7bade986.png)

pikurasa commented:
I made a test file and I do hear a difference between equal and just. (apologies for the false alarm -- I needed to make notes sound longer to hear the difference. That being said, I do think the idea for block for status widget would help)

I am unsure what to listen for with meantone as I am less experienced with it, but will do some research.

Here is the test file:
[Temperament-Test.html.zip](https://github.com/sugarlabs/musicblocks/files/2023916/Temperament-Test.html.zip)

pikurasa commented:
@riyalohia 

There are two common types of meantone, 1/4 comma and 1/3 comma. I think that both of those types should be available as a choice in the default dropdown.

https://en.wikipedia.org/wiki/Meantone_temperament

pikurasa commented:
...hmmm...

I have been trying different notes for the chords and it does not seem to be playing all of the notes.

This is a screenshot of me trying e flat. The note is going to the log, but not being performed.
![screenshot at 2018-05-21 16 24 40](https://user-images.githubusercontent.com/13454579/40328206-99b478d4-5d13-11e8-9fd4-da4d0209134c.png)

pikurasa commented:
A little more testing and I see that if I rearrange the notes, I will hear them.

I tried moving the eflat to the bottom of the stack, and it performed.

** Something that needs some investigating is the sound seems to perform (from a previous run of the code, maybe?) before the current code has been run. The behavoir is suspicious. I will try and make a video.

![screenshot at 2018-05-21 16 27 09 bottom stack](https://user-images.githubusercontent.com/13454579/40328307-eb4658fc-5d13-11e8-9732-46d562a21050.png)

pikurasa commented:
> ** Something that needs some investigating is the sound seems to perform (from a previous run of the code, maybe?) before the current code has been run. The behavoir is suspicious. I will try and make a video.

I am having trouble getting the sound to be properly recorded, but this video shows the steps to reproduce the issue.

[2018-05-21-2018-temperament-2.ogv.zip](https://github.com/sugarlabs/musicblocks/files/2024017/2018-05-21-2018-temperament-2.ogv.zip)

pikurasa commented:
The audio for this video works.

Notice that the sound starts twice when pressing "run"

[2018-05-21-2018-temperament-3.ogv.zip](https://github.com/sugarlabs/musicblocks/files/2024028/2018-05-21-2018-temperament-3.ogv.zip)

walterbender commented:
@riyalohia  Let's start a separate PR for the widget.

walterbender commented on js/basicblocks.js:
```diff
@@ -1123,6 +1123,14 @@ function initBasicProtoBlocks(palettes, blocks) {
+    var pitchInHertzBlock = new ProtoBlock('pitchinhertz');
+    pitchInHertzBlock.palette = palettes.dict['meter'];
+    blocks.protoBlockDict['pitchinhertz'] = pitchInHertzBlock;
+    //.TRANS: number of whole notes that have been played
```
This translator note should say: the current pitch expressed in Hertz

walterbender commented on js/logo.js:
```diff
@@ -9324,6 +9324,11 @@ function Logo () {
+            case 'pitchinhertz':
+                if (that.inStatusMatrix && that.blocks.blockList[that.blocks.blockList[blk].connections[0]].name === 'print') {
+                    that.statusFields.push([blk, 'pitchinhertz']);
+                }
```
else {
    that.blocks.blockList[blk].value = ...
}

walterbender replied:
I think it should return the frequency in Hertz of the note being played. (You can only return on value, since it can be a chord, so just return the first entry in the array.)

walterbender commented:
Regarding the temperament name block, please look at:
blocks.js and palette.js for how to add the secret sauce needed for selector blocks (like voicename, notename, etc.) Something similar needs to be added for temperament name.

e.g.,
https://github.com/sugarlabs/musicblocks/blob/master/js/blocks.js#L2157
https://github.com/sugarlabs/musicblocks/blob/master/js/blocks.js#L3883
https://github.com/sugarlabs/musicblocks/blob/master/js/palette.js#L738

Also, I think it would be good to define the temperaments for the selector in utils/musicutils.js, along with a default as per the other selectors.

Finally, due to a bug in the translation code, you'll need to add the selector strings to:
turtledefs.js

riyalohia commented:
@walterbender , As per your suggestion, I have made changes in blocks.js and palette.js, but I am not really sure whether I have done the correct thing ?

walterbender commented on js/utils/musicutils.js:
```diff
@@ -373,6 +373,20 @@ var OSCTYPES = [
     [_('sawtooth'), 'sawtooth'],
 ];
 
+var TEMPERAMENTS = [
```
This is good. You should use these definitions in block.js
